File: scala/com/kotori316/transformer/gui/GuiSource.scala
package com.kotori316.transformer.gui

import com.kotori316.transformer.EnergyTranformer
import com.kotori316.transformer.block.TileSource
import com.kotori316.transformer.network.{PacketHandler, SourceAmountMessage}
import net.minecraft.client.gui.GuiTextField
import net.minecraft.client.gui.inventory.GuiContainer
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.ResourceLocation
import org.apache.commons.lang3.math.NumberUtils
import org.lwjgl.input.Keyboard
import org.lwjgl.opengl.GL11

class GuiSource(player: EntityPlayer, tile: TileSource) extends GuiContainer(new ContainerSource(player, tile)) {
    val LOCATION = new ResourceLocation(EnergyTranformer.modID, "textures/gui/energytranformer.png")
    var text: GuiTextField = _

    override def initGui(): Unit = {
        super.initGui()
        Keyboard.enableRepeatEvents(true)
        text = new GuiTextField(0, this.fontRenderer, getGuiLeft + 18, getGuiTop + 18, 50, 10)
        text.setText(tile.amount.toString)
    }

    override def onGuiClosed(): Unit = {
        super.onGuiClosed()
        Keyboard.enableRepeatEvents(false)
        if (NumberUtils.isDigits(text.getText)) {
            val i = text.getText.toInt
            if (i > 0) {
                tile.amount = i
                PacketHandler.INSTANCE.sendToServer(new SourceAmountMessage(i, tile.getPos))
            }
        }
    }

    override def keyTyped(typedChar: Char, keyCode: Int): Unit = {
        text.textboxKeyTyped(typedChar, keyCode)
        super.keyTyped(typedChar, keyCode)
    }

    override def mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int): Unit = {
        text.mouseClicked(mouseX, mouseY, mouseButton)
        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override def drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float): Unit = {
        this.drawDefaultBackground()
        super.drawScreen(mouseX, mouseY, partialTicks)
        this.renderHoveredToolTip(mouseX, mouseY)
        text.drawTextBox()
    }

    override def drawGuiContainerBackgroundLayer(partialTicks: Float, mouseX: Int, mouseY: Int) = {
        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F)
        this.mc.getTextureManager.bindTexture(LOCATION)
        this.drawTexturedModalRect(guiLeft, guiTop, 0, 0, xSize, ySize)
    }

    override def drawGuiContainerForegroundLayer(mouseX: Int, mouseY: Int): Unit = {
        super.drawGuiContainerForegroundLayer(mouseX, mouseY)
    }

}

File: src/main/java/com/kotori316/transformer/gui/GuiSource.java
package com.kotori316.transformer.gui;

import net.minecraft.client.gui.GuiTextField;
import net.minecraft.client.gui.IGuiEventListener;
import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.StringUtils;
import org.apache.commons.lang3.math.NumberUtils;

import com.kotori316.transformer.EnergyTransformer;
import com.kotori316.transformer.TileSource;

public class GuiSource extends GuiContainer {
    private static final ResourceLocation LOCATION = new ResourceLocation(EnergyTransformer.modID, "textures/gui/" + EnergyTransformer.modID + ".png");
    private final TileSource tileSource;
    private GuiTextField text;

    public GuiSource(EntityPlayer player, TileSource tileSource) {
        super(new ContainerSource(player));
        this.tileSource = tileSource;
    }

    @Override
    public IGuiEventListener getFocused() {
        return this.text.isFocused() ? this.text : null;
    }

    @Override
    protected void initGui() {
        super.initGui();
        mc.keyboardListener.enableRepeatEvents(true);
        text = new GuiTextField(0, this.fontRenderer, getGuiLeft() + 18, getGuiTop() + 18, 50, 10);
        text.setText(String.valueOf(tileSource.getStorage().getMaxEnergyStored()));
        text.setTextAcceptHandler(this::updateCapacity);
        text.setValidator(this::predicateString);
        this.children.add(text);
    }

    @Override
    public void onGuiClosed() {
        super.onGuiClosed();
        this.mc.keyboardListener.enableRepeatEvents(false);
    }

    @Override
    public void render(int mouseX, int mouseY, float partialTicks) {
        this.drawDefaultBackground();
        super.render(mouseX, mouseY, partialTicks);
        this.renderHoveredToolTip(mouseX, mouseY);
        GlStateManager.disableLighting();
        GlStateManager.disableBlend();
        this.text.drawTextField(mouseX, mouseY, partialTicks);
    }

    @Override
    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY) {
        GlStateManager.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        this.mc.getTextureManager().bindTexture(LOCATION);
        this.drawTexturedModalRect(guiLeft, guiTop, 0, 0, xSize, ySize);
    }

    private void updateCapacity(int id, String text) {
        final int newCapacity;
        if (StringUtils.isNullOrEmpty(text)) {
            newCapacity = 0;
        } else {
            int i;
            try {
                i = Integer.parseInt(text);
            } catch (NumberFormatException ignore) {
                i = 0;
            }
            newCapacity = i;
        }
        tileSource.updateCapacity(newCapacity);
    }

    private boolean predicateString(String text) {
        return StringUtils.isNullOrEmpty(text) || NumberUtils.isDigits(text);
    }
}
